### ISSUE TITLE
fine-tune time and decode part

### ISSUE THREAD

**Issue by Brandonnogithub** [microsoft/unilm#40]
Hello, I try to use your model on QG task. However it takes a lot of time to fine-tune. 
It takes 8h for each epoch on SQuAD dataset. (no fp16 since my GPU is not suppoted)
The GPU is 1080ti (11g) and I need to set the batch size to 1.

I wonder why this model is so slow in fine-tune part ? (compare to other pre-train model like GPT and BERT)

Another question is about the decode part. Each time predict a word and feed this updated sequence back to model to predict next word. Is this right? (Sorry the codes are too complex to understand for me.)

**Comment by nyanyanya** [comment]
HI Brandon,

The fine-tune cost/speed should be similar to BERT model of the same size. Since you are not using fp-16, a lot of GPU memory will be used to store the parameters in the model and the optimizer. I would still use fp-16 in training to save memory and to enable larger batch size,  even if your GPU does not have special support for fp-16 computation like tensor-core.

The decoding is conceptually very simple: for a sequence [y(0), ..., y(n-1)], we feed [y(0), ..., y(n-1), mask] into the model to get the prediction of y(n). It can be done by brute-forcefully compute all layers for all n input tokens, but certainly is not very efficient. Instead, we want to re-use the computed hidden representations from previous step. But we can only get hidden representations for [y(0), ..., y(n-2)] from previous step, because the input for previous step was [y(0), ..., y(n-2), mask] and y(n-1) was not predicted yet. As a result, in every step, we need to compute the representation for two tokens [y(n-1), mask] while re-using representations of [y(0), ..., y(n-2)]. This is the major difference between our decoder and the usual decoder which does not use masks.

**Comment by Brandonnogithub** [comment]
Thank you. I got it. I will try this model on cloud gpu.

**Comment by liebkne** [comment]
@nyanyanya Hi. To get `y(n)`, why `[y(0), ..., y(n-1), mask]` is needed instead of `[y(0), ..., y(n-1)]`? How is the decoding in UniLM different from a normal RNN Seq2seq decoder?
